src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+///           - The number of UnitTypeStats is known at start + a properly sized vector =
+///             simple pointers to UnitTypeStats can be used safely
+///           - Huge reduction of the code base.
+///           - Cleaner implemention of the gamelogic.
+///           - Simpler information retrieval for the gui.
+///           - Performance (size, locality, less indirection)
+///
+///
+struct CivilizationStats {
+
+  // Add to the starting resource of the player
+  ResourceAmount startingResourceD;
+
+  UnitType startingScoutUnit;
+  int startingVillagerCount;
+  bool startingLlama;
+
+  int populationCap;
+  int bonusPopulationSpace;
+
+  float garrisonHealRate;
+
+  // TODO: monk conversion things
+
+  float tributeFee;
+  float tradingFree;
+  float tradeGoldGeneration;
+  float relicGoldGeneration;
+
+  // == single unit values
+
+  int farmFood;
+
+  // NOTE: could aslo be stored as the workRate of an unspecialized villager
+  float villagerConstuctionSpeed;
+
+  float villagerCarryingCapacity[/* TODO: number of villager types*/];
+  float fishingShipCarryingCapacity;
+
+  bool singleTowerCenter;
+  // use singleTowerCenter this instead of a max limit for every building (only the town center has
+  // max limit, even the wonder can be build multiple types)
+
+};
+
+/// Represents a modification to something
+struct Modification {
+
+  static constexpr i32 Unset = std::numeric_limits<i32>::min();
+
+  enum class Type {
+    // unit, building and technology
+    Cost,            // use extra to define ResourceType or leave Unset for to all of them
+
+    // unit and building
+    MaxHp,
+    Damage,          // use extra to define DamageType
+    Armor,           // use extra to define DamageType
+    MinRange,
+    MaxRange,
+    FireRate,
+    Accuracy,        // use extra to define type of accuracy ?
+    LineOfSight,
+    GarrisonCapacity,
+    Upgrade,         // option 1: use value to define the old type, use extra to define the new unit (applied to civ)
+                     // option 2: use value to define the new unit (applied to unit)
+
+    // unit only
+    Speed,
+    ProductionTime,
+
+    // building only
+    ConstructionSpeed,
+    WorkRate,
+    PopulationSpace,
+
+    // technology only
+    ResearchTime,
+
+    // civilization only
+    VillagerCarryingCapacity,  // use extra to define villager type or leave Unset for all of them
+    PopulationCap
+  };
+
+  // TODO: combine ProductionTime, ConstructionSpeed and ResearchTime ?
+
+  // NOTE: Not all civilization stats need to have a corresponding modification type.
+  //       Eg. the startingScoutUnit can just be set at the begging of the game and cannot be changed later.
+
+  // TODO: find a clever way to compress all the monk conversion things
+
+  enum class Operation {
+    Set, // Set the value
+    Add, // Add the value
+    Mult, // Multiply the value
+    MultAdd, // Multiply the base value and add the result to the value
```

Comment:
I added the Mult first, then I realized that something like MultAdd is needed for all the % changes to the values. However now I dont think that the Mult is needed for anything (it's also not a good game design because introduces results that the player cannot understand eg: simpler `+20% and +10% = 30%` than `+20% and +10% = 32%`).

So I would suggest (if we not find anything that needs the Mult) to remove it and rename the MultAdd to Mult.

Yes, the comments are a mess.